consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,165 @@
+        let order = FairnessShufflerImpl::new(self, &conflict_key_registries).shuffle();
+        reorder(txns, order)
+    }
+}
+
+fn reorder(txns: Vec<SignedTransaction>, order: Vec<TxnIdx>) -> Vec<SignedTransaction> {
+    assert_eq!(txns.len(), order.len());
+    order.into_iter().map(|idx| txns[idx].clone()).collect()
+}
+
+struct FairnessShufflerImpl<'a> {
+    conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    conflict_zones: [ConflictZone<'a>; NUM_CONFLICT_ZONES],
+    pending_zones: [PendingZone<'a>; NUM_CONFLICT_ZONES],
+    selected_order: Vec<TxnIdx>,
+    selection_tracker: SelectionTracker,
+}
+
+impl<'a> FairnessShufflerImpl<'a> {
+    pub fn new(
+        shuffler: &FairnessShuffler,
+        conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    ) -> Self {
+        let num_txns = conflict_key_registries[0].num_txns();
+        assert!(conflict_key_registries
+            .iter()
+            .skip(1)
+            .all(|r| r.num_txns() == num_txns));
+
+        Self {
+            conflict_key_registries,
+            selected_order: Vec::with_capacity(num_txns),
+            selection_tracker: SelectionTracker::new(num_txns),
+            conflict_zones: ConflictZone::build_zones(conflict_key_registries, [
+                shuffler.sender_conflict_window_size,
+                shuffler.module_conflict_window_size,
+                shuffler.entry_fun_conflict_window_size,
+            ]),
+            pending_zones: PendingZone::build_zones(conflict_key_registries),
+        }
+    }
+
+    pub fn shuffle(mut self) -> Vec<TxnIdx> {
+        // First pass, only select transactions with no conflicts in all conflict zones
+        while let Some(txn_idx) = self.selection_tracker.next_unselected() {
+            if !self.is_conflict(txn_idx) && !self.is_hol_blocked(txn_idx) {
+                self.select_and_select_unconflicted(txn_idx, false /* is_pending */)
+            } else {
+                self.append_pending(txn_idx);
+            }
+        }
+
+        // Second pass, select previously pending txns in order,
+        //   with newly un-conflicted txns jumping the line
+        self.selection_tracker.new_pass();
+        while let Some(txn_idx) = self.selection_tracker.next_unselected() {
+            self.select_and_select_unconflicted(txn_idx, true /* is_pending */);
+        }
+
+        self.selected_order
+    }
+
+    fn select_and_select_unconflicted(&mut self, txn_idx: TxnIdx, is_pending: bool) {
+        let mut maybe_unconflicted = self.select(txn_idx, is_pending);
+        while let Some(txn_idx) = maybe_unconflicted.pop_first() {
+            if !self.is_conflict(txn_idx) && !self.is_hol_blocked(txn_idx) {
+                maybe_unconflicted.extend(self.select(txn_idx, true /* is_pending */))
+            }
+        }
+    }
+
+    // Select a transaction and return potentially un-confilcted transactions
```

Comment:
nits:
* typo: confilcted -> conflicted
* can `///` we used so the IDE picks up the function doc? :)